hazelcast/src/main/java/com/hazelcast/internal/probing/ProbeRegistry.java

```diff
@@ -0,0 +1,107 @@
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.spi.annotation.Beta;
+import com.hazelcast.spi.annotation.PrivateApi;
+import com.hazelcast.spi.impl.servicemanager.ServiceManager;
+
+/**
+ * A service made accessible to core services so they have a chance to register
+ * the "root objects" of the system.
+ * 
+ * After the registration phase the set of this {@link ProbeSource}s is fix.
+ * There is no need to unregister them later on as they always have the choice
+ * to stop rendering something.
+ * 
+ * Metrics about things that change over time require a corresponding
+ * {@link ProbeSource} (existing since the startup) that does probe the
+ * available instances as they appear and stops doing that as they disappear.
+ */
+@Beta
+@PrivateApi
+public interface ProbeRegistry {
+
+    /**
+     * Called once at startup, typically by a core service registering itself.
+     * 
+     * @param source a object that "knows" how to render metrics in its context
+     */
+    void register(ProbeSource source);
+
+    /**
+     * Legacy support instances only known by their interface with implementations
+     * that possibly implement {@link ProbeSource}.
+     * 
+     * @param source a object possibly implementing {@link ProbeSource}, might be
+     *        null as well
+     */
+    void registerIfSource(Object source);
+
+    /**
+     * @return Creates a new "private "context that should be kept by the caller to
+     *         render the contents of this {@link ProbeRegistry}. The implementation
+     *         will not support multi-threading as each thread should create its own
+     *         context instance.
+     */
+    ProbeRenderContext newRenderingContext();
+
+    ProbeRenderContext newRenderingContext(Class<? extends ProbeSource>... filter);
+
+    /**
+     * From a usability point of view the {@link ProbeRenderContext} is a bit
+     * cumbersome and smells like over-abstraction. It is purely introduced to
+     * achieve the goal of rendering without creating garbage objects. That means
+     * state needs to be reused. This object is the place where state can be kept in
+     * a way that allows reuse between rendering cycles.
+     * 
+     * The {@link ProbeRenderer} itself usually changes for each cycle as it tends
+     * to be dependent on output stream objects handed to it.
+     */
+    interface ProbeRenderContext {
+
+        /**
+         * Causes a {@link ProbingCycle} that is directed at the given
+         * {@link ProbeRenderer}.
+         * 
+         * This method does not support multi-threading. If potentially concurrent calls
+         * to this method should be made each should originate from its own
+         * {@link ProbeRenderContext}.
+         * 
+         * @param renderer not null; is called for each active prove with a key and
+         *        value to convert them to the renderer specific format.
+         */
+        void renderAt(ProbeLevel level, ProbeRenderer renderer);
```

Comment:
Why not call this method 'render'? Passing the probe level already implies that the probe level is relevant. The 'at' part doesn't provide more information.

Reply:
Fine. I thought this reads nicer and also helps not confuse the two `render` methods in different interfaces.